sbin/pppconfig/pppconfig.8
.\" Copyright (c) 1993 Berkeley Software Design, Inc. All rights reserved.
.\" The Berkeley Software Design Inc. software License Agreement specifies
.\" the terms and conditions for redistribution.
.\"	BSDI $Id: pppconfig.8,v 1.1.1.1 1993/03/09 15:17:53 polk Exp $
.Dd February 24, 1993
.Dt PPPCONFIG 8
.Os
.Sh NAME
.Nm pppconfig
.Nd configure Point-to-Point Protocol parameters
.Sh SYNOPSIS
.Nm pppconfig
.Ar interface
.Op Ar parameters
.Sh DESCRIPTION
.Nm Pppconfig
is used to examine and/or set parameters of a point-to-point
protocol link.
When called without
.Ar parameters ,
.Nm pppconfig
will print up to three sets of parameters:
the initial parameters as used before negotiating
the PPP options, the current negotiated parameters if
the link is active,
and the non-negotiated parameters which
do not require the negotiation between the local and remote side.
.Pp
.Nm Pppconfig
can be used to change the initial and
non-negotiated parameters at any time (superuser permissions
required);
however it is not possible to change the negotiated parameters
.Dq "on the fly"
\- the connection should be terminated and
reestablished for the new parameters to take effect.
If the link is not active the negotiated parameters
assume the initial values.
.Pp
The
.Ar interface
parameter is a string of the form
.Dq name unit ,
for example,
.Dq rn0 .
.Pp
The following parameters may be set with 
.Nm pppconfig :
.Bl -tag -width cmap_charmap
.It Cm cmap Ar charmap
Specify which control characters should be avoided on
asynchronous lines in case the equipment uses some
of them for flow control or link control or inserts
characters for time fill.
The
.Ar charmap
may be specified as a 32-bit hexadecimal or octal number
(if it starts with 0x or 0 respectively), in which
case the least significant bit corresponds to the null
character and the most significant bit to the character
with code 037 (US).
Alternatively, it may be specified as a string of
letters corresponding the control characters, for example
.Dq @QS
will represent mask with bits 0, 17 and 19 set.
The default all-zero mask will allow transmission of
all control characters.
.It Cm mru Ar size
Specify the maximal packet size the local side is willing
to receive.
This option can be useful on packet-switching links
to prevent link-level packet fragmentation or in case
of problems with hardware flow control.
The default value for
.Cm mru
is 1500 bytes.
.It Cm pfc
Enable protocol field compression.
This option eliminates an extra byte in PPP packet header,
which can be useful on slow links but can cause computing
overhead on fast synchronous links.
This option is default on asynchronous lines.
.It Fl pfc
Disable protocol field compression.
This option is default on synchronous lines.
.It Cm acfc
Enable address and control fields compression.
This option reduces the size of a PPP packet header by two bytes,
which can be useful on slow links but can cause computing
overhead on fast synchronous links.
This option is default on asynchronous lines.
.It Fl acfc
Disable address and control fields compression.
This option is default on synchronous lines.
.It Cm tcpc
Enable Van Jacobson's TCP header compression.
This option enables compression of TCP headers by
eliminating redundant information in the headers
of consecutive TCP packets, reducing the protocol
overhead by up to 100 bytes per packet, which may be
really critical to the performance of interactive
.Nm telnet
or
.Nm rlogin
sessions.
However, VJ TCP compression requires the system to keep
state information about every TCP session on both
ends (up to 16 simultaneous sessions) and is not
really useful on fast links connecting large networks.
TCP compression does not affect UDP and ICMP packets.
This option is default on asynchronous lines.
.It Fl tcpc
Disable Van Jacobson's TCP header compression.
This option is default on synchronous lines.
.It Cm ftel
Enable high-priority queueing of
.Dq interactive
TCP packets
(with source or destination ports assigned to
services like
.Xr telnet
or control connection of
.Xr ftp 1 ) ,
effectively giving them precedence over all other IP packets.
This can reduce response time on links with mixed
file transfer and interactive traffic.
This function is a temporary and limited substitute for the
still unsupported IP low delay TOS (type of service).
This option is default on asynchronous lines.
.It Fl ftel
Disable the
.Dq "fast telnet"
option.
This is default on synchronous lines.
.It Cm trace
Enable printing of trace information for PPP control packets and protocol
finite state machine transitions.
This option is effective only if the kernel is configured to
include the PPP trace printing code.
The default value for this option is defined by the kernel
configuration.
.It Fl trace
Disable printing of trace information for PPP control packets and protocol
finite state machine transitions.
.It Cm idletime Ar n
Set the idle timer to drop the connection if no outgoing
data packets were processed during the last
.Ar n
seconds (the maximal time is 32767 seconds).
.It Fl idletime
Disable the idle timer.
This option is default.
.It Cm maxconf Ar n
Set the limit on attempts to negotiate PPP control protocol
options to
.Ar n .
After the specified number of unsuccessful attempts the connection
will be dropped.
This option is useful on dial-up links to force the connection to drop
when there is a malfunction on the remote side.
The default value on asynchronous lines is 10.
The maximal value of
.Cm maxconf
is 255.
.It Fl maxconf
Allow unlimited attempts to negotiate PPP options.
This option should be used on dedicated lines and is default
on synchronous lines.
.It Cm maxterm Ar n
Set the limit on attempts to terminate the link gracefully
to
.Ar n.
The default is 3 attempts before unilaterally dropping the
connection.
The maximal value of
.Cm maxterm
is 255.
.It Cm timeout Ar n
Specify the interval between attempts to configure/terminate
the link in tenths of a second.
The default value is 30 (3 seconds), the maximal value is 255
(25.5 seconds).
.El
.Pp
Only the super-user may modify the configuration of a network interface.
.Pp
If the interface supports more than one link-level protocol
(for example, a synchronous link), the PPP protocol should be selected before
configuring the PPP parameters (see
.Xr ppp 4
and
.Xr ifconfig 8 ) .
The default values will be loaded when the link-level protocol
is switched to PPP from some other protocol.
.Sh DIAGNOSTICS
Messages indicating the specified interface does not exist, the
requested address is unknown, or the user is not privileged and
tried to alter an interface's PPP configuration.
.Sh SEE ALSO
.Xr ifconfig 8 ,
.Xr ppp 4 ,
.Xr pppattach 8 ,
.Xr ppp 8 ,
.Xr rn 4 ,
.Xr rh 4 ,
.Xr appp 4
.br
.Pp
W. Simpson,
The Point-to-Point Protocol (PPP) for the Transmission of
Multi-protocol Datagrams over Point-to-Point Links, RFC\ 1331,
May 1992.
.Pp
G. McGregor,
The PPP Internet Protocol Control Protocol (IPCP), RFC\ 1332,
May 1992.
.Sh HISTORY
The
.Nm pppconfig
command is
.Ud .


sbin/pppconfig/pppconfig.c
/*
 * Copyright (c) 1993 Berkeley Software Design, Inc. All rights reserved.
 * The Berkeley Software Design Inc. software License Agreement specifies
 * the terms and conditions for redistribution.
 */

static const char rcsid[] = "$Id: pppconfig.c,v 1.2 1994/01/28 00:11:26 sanders Exp $";
static const char copyright[] = "Copyright (c) 1993 Berkeley Software Design, Inc.";

/*
 * Copyright (c) 1983 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/* The skeleton code literally repeats ifconfig */

#include <sys/param.h>
#include <sys/socket.h>
#include <sys/ioctl.h>

#include <net/if.h>

#include <unistd.h>
#include <stdio.h>
#include <errno.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>

#include <net/pppioctl.h>

struct  ifreq     ifr, ifr1;

struct ppp_ioctl *pio  = ifr_pppioctl(&ifr);
struct ppp_ioctl *pio1 = ifr_pppioctl(&ifr1);

int chaged;
char	name[30];

int setflags(), setcmap(), mru(), idletime(), maxconf(), maxterm(), timeout();

#define	NEXTARG		0xffffff
#define ZERO            0xfffffe

struct	cmd {
	char	*c_name;
	int	c_parameter;		/* NEXTARG means next argv */
	int	(*c_func)();
} cmds[] = {
	/* cmap */
	{ "cmap",       NEXTARG,        setcmap },

	/* mru */
	{ "mru",        NEXTARG,        mru },

	/* idletime */
	{ "idletime",   NEXTARG,        idletime },
	{ "-idletime",  ZERO,           idletime },

	/* flags */
	{ "pfc",        PPP_PFC,        setflags },
	{ "-pfc",       -PPP_PFC,       setflags },
	{ "acfc",       PPP_ACFC,       setflags },
	{ "-acfc",      -PPP_ACFC,      setflags },
	{ "tcpc",       PPP_TCPC,       setflags },
	{ "-tcpc",      -PPP_TCPC,      setflags },
	{ "ftel",       PPP_FTEL,       setflags },
	{ "-ftel",      -PPP_FTEL,      setflags },
	{ "trace",      PPP_TRACE,      setflags },
	{ "-trace",     -PPP_TRACE,     setflags },

	/* maxconf */
	{ "maxconf",    NEXTARG,        maxconf },
	{ "-maxconf",   ZERO,           maxconf },

	/* maxterm */
	{ "maxterm",    NEXTARG,        maxterm },

	/* "timeout" */
	{ "timeout",    NEXTARG,        timeout },

	{ 0,            0,              0 }
};

main(argc, argv)
	int argc;
	char *argv[];
{
	int s;

	if (argc < 2) {
		fprintf(stderr, "usage: pppconfig interface\n%s\n%s\n%s\n",
			"\t[ cmap hex-map ] [ mru n ] [ idletime secs | -idletime ]",
			"\t[ pfc | -pfc ] [ acfc | -acfc ] [ tcpc | -tcpc ] [ ftel | -ftel ]",
			"\t[ maxconf n | -maxconf ] [ maxterm n ] [ timeout 1/10secs ]");
		exit(1);
	}
	argc--, argv++;
	strncpy(name, *argv, sizeof(name));
	strncpy(ifr.ifr_name, name, sizeof ifr.ifr_name);
	argc--, argv++;
	s = socket(AF_INET, SOCK_DGRAM, 0);
	if (s < 0) {
		perror("ifconfig: socket");
		exit(1);
	}
	if (ioctl(s, PPPIOCGPAR, (caddr_t)&ifr) < 0) {
		Perror("ioctl (PPPIOCGPAR)");
		exit(1);
	}

	if (argc == 0) {
		strncpy(ifr1.ifr_name, name, sizeof ifr1.ifr_name);
		if (ioctl(s, PPPIOCNPAR, (caddr_t)&ifr1) < 0) {
			Perror("ioctl (PPPIOCGPAR)");
			exit(1);
		}

		printf("initial:\tcmap=0x%x, mru=%d, %spfc, %sacfc, %stcpc\n",
			pio->ppp_cmap, pio->ppp_mru,
			(pio->ppp_flags & PPP_PFC) ? "" : "-",
			(pio->ppp_flags & PPP_ACFC) ? "" : "-",
			(pio->ppp_flags & PPP_TCPC) ? "" : "-");
		printf("negotiated:\tcmap=0x%x, mru=%d, %spfc, %sacfc, %stcpc\n",
			pio1->ppp_cmap, pio1->ppp_mru,
			(pio1->ppp_flags & PPP_PFC) ? "" : "-",
			(pio1->ppp_flags & PPP_ACFC) ? "" : "-",
			(pio1->ppp_flags & PPP_TCPC) ? "" : "-");
		printf("non-negotiable:\t");
		if (pio->ppp_flags & PPP_FTEL)
			printf("ftel,");
		else
			printf("-ftel,");
		if (pio->ppp_flags & PPP_TRACE)
			printf(" trace,");
		else
			printf(" -trace,");
		if (pio->ppp_idletime == 0)
			printf(" -idletime,");
		else
			printf(" idletime=%d,", pio->ppp_idletime);
		printf("\n\t\t");
		if (pio->ppp_maxconf == 0)
			printf("-maxconf,");
		else
			printf("maxconf=%d,", pio->ppp_maxconf);
		printf(" maxterm=%d, timeout=%d\n",
		       pio->ppp_maxterm, pio->ppp_timeout);
		exit(0);
	}

	*pio1 = *pio;
	while (argc > 0) {
		register struct cmd *p;

		for (p = cmds; p->c_name; p++)
			if (strcmp(*argv, p->c_name) == 0)
				break;
		if (p->c_name == 0) {
			fprintf(stderr, "pppconfig: illegal parameter name %s\n", *argv);
			exit(1);
		}
		if (p->c_func) {
			if (p->c_parameter == NEXTARG) {
				(*p->c_func)(argv[1]);
				argc--, argv++;
			} else if (p->c_parameter == ZERO)
				(*p->c_func)("0");
			else
				(*p->c_func)(p->c_parameter);
		}
		argc--, argv++;
	}
	strncpy(ifr.ifr_name, name, sizeof ifr.ifr_name);
	if (memcmp(pio, pio1, sizeof *pio) &&
	    ioctl(s, PPPIOCSPAR, (caddr_t)&ifr) < 0) {
		Perror("ioctl (PPPIOCSPAR)");
		exit(1);
	}
	exit(0);
}

setflags(val)
	int val;
{

	if (val < 0)
		pio->ppp_flags &= ~(-val);
	else
		pio->ppp_flags |= val;
}

setcmap(arg)
	char *arg;
{
	u_long map = 0;
	char c;

	if (*arg == '0') {
		/* Hexadecimal? */
		if (*++arg = 'x' || *arg == 'X') {
			while (c = *++arg) {
				map <<= 4;
				if ('0' <= c && c <= '9')
					map |= c - '0';
				else if ('A' <= c && c <= 'F')
					map |= c - 'A' + 10;
				else if ('a' <= c && c <= 'f')
					map |= c - 'a' + 10;
				else
					goto err;
			}
		} else {
			/* Octal? */
			while (c = *arg++) {
				map <<= 3;
				if ('0' <= c && c <= '7')
					map |= c - '0';
				else
					goto err;
			}
		}
	} else {
		/* List of characters? */
		while (c = *arg++)
			map |= 1 << (c & 037);
	}
	pio->ppp_cmap = map;
	return;
err:
	fprintf(stderr, "pppconfig: bad character map\n");
	exit(1);
}

mru(arg)
	char *arg;
{
	int v = atoi(arg);

	if (v < 128 || v > 1500) {
		fprintf(stderr, "pppconf: bad value for mru -- must be from 128 to 1500\n");
		exit(1);
	}
	pio->ppp_mru = v;
}

idletime(arg)
	char *arg;
{

	pio->ppp_idletime = atoi(arg);
}

maxconf(arg)
	char *arg;
{

	pio->ppp_maxconf = atoi(arg);
}

maxterm(arg)
	char *arg;
{

	if ((pio->ppp_maxterm = atoi(arg)) == 0) {
		fprintf(stderr, "pppconf: bad value for maxterm -- should be non-zero\n");
		exit(1);
	}
}

timeout(arg)
	char *arg;
{

	if ((pio->ppp_timeout = atoi(arg)) == 0) {
		fprintf(stderr, "pppconf: bad value for timeout -- should be non-zero\n");
		exit(1);
	}
}

Perror(cmd)
	char *cmd;
{
	extern int errno;

	fprintf(stderr, "pppconfig: ");
	switch (errno) {

	case ENXIO:
		fprintf(stderr, "%s: no such interface\n", cmd);
		break;

	case EPERM:
		fprintf(stderr, "%s: permission denied\n", cmd);
		break;

	default:
		perror(cmd);
	}
	exit(1);
}
